Fetch billing accounts/projects from Google

breilly2 opened the pull request:
Resolves: #1275 
Supports: #1188 

This PR does not contain any user-facing changes, only utilities to help with billing UI.

Usage example:
```
    await ensureBillingScope()
    const accountNames = _.map('name', await GoogleBilling.listAccounts()) // or fetch account names from FireCloud
    const projectNames = await Promise.all(_.map(async accountName => GoogleBilling.listProjectNames(accountName), accountNames))
    const projectsByAccount = _.fromPairs(_.zip(accountNames, projectNames))
```
~~withBillingScope(async () => {
  const accounts = await GoogleBilling.listAccounts()
  const projects = await _.flow(
    _.map('name'),
    GoogleBilling.listProjectsWithAccounts,
  )(accounts)
  doSomethingWith(projects)
}, () => {
  reportError('Google Cloud Billing scope not granted')
})~~

breilly2 committed:
Added functions for fetching billing accounts and their associated projects from Google.
Added utilities for wrapping code that needs cloud billing scope.

breilly2 commented:
This PR is originally submitted against dev. Another thought I had was to submit it against Kate's branch so she'll get the changes. Either way is fine with me.

breilly2 commented on src/libs/ajax.js:
```diff
@@ -655,6 +658,39 @@ const Buckets = signal => ({
+  },
+
+  listProjectsWithAccounts: async billingAccountNames => {
+    return _.fromPairs(
```
This works, but it feels like an obvious place where `_.flow()` would work well. I'm going to experiment with that.

breilly2 commented on src/libs/auth.js:
```diff
@@ -21,6 +21,28 @@ export const reloadAuthToken = () => {
+ * In this case, you'll need to provide something for the user to deliberately click on and retry
+ * withBillingScope in reaction to the click.
+ */
+export const withBillingScope = (f, onFailure) => {
```
My Scala is showing a little bit here. Let me know if there's a better way to express this pattern.

panentheos replied:
This could just be an async function, e.g. `ensureBillingScope`. If we have the scope already, just return. Otherwise, await on the google stuff, and then return. Then the consumer can just do `await ensureBillingScope()`

zarsky-broad commented on src/libs/ajax.js:
```diff
@@ -655,6 +658,39 @@ const Buckets = signal => ({
+          _.map(
+            async name => fetchGoogleBilling(`${name}/projects`, _.merge(authOpts(), { signal })).then(
+              response => response.json().then(
+                json => [name, _.map(
```
```suggestion
                json => _.map('projectId', json.projectBillingInfo)
```

breilly2 committed:
Refactor

elizzhao commented on src/libs/ajax.js:
```diff
@@ -655,6 +658,34 @@ const Buckets = signal => ({
+      // This might happen if the user has permission to create projects with this account
+      // but is missing the billing.resourceAssociations.list permission.
+      if (errorResponse.status === 403) {
+        return []
```
Should we return something more descriptive in this error state?

breilly2 replied:
This isn't really an error state. This translates the `403 Forbidden` response (due to lack of permissions) to "no data". The comment above is intended to communicate that. I can make that more explicit. I was intentionally a little bit aloof about it because I don't know what leads to this state, or if its even really a valid state to be in or just something I came across with dev data. What I do know is that I want to do this translation.

However, maybe the "I" in that statement is the caller of this function rather than the provider of this API. Moving this out to the caller would give the caller more flexibility about how to interpret this result, which might turn out to be important. I'll make that change unless anyone objects.

elizzhao commented on src/libs/auth.js:
```diff
@@ -21,6 +21,28 @@ export const reloadAuthToken = () => {
+  } else {
+    const options = new window.gapi.auth2.SigninOptionsBuilder({ 'scope': 'https://www.googleapis.com/auth/cloud-billing' })
+    // Wait 100ms before doing the thing to avoid errors due to delays in applying the new scope grant
+    getAuthInstance().currentUser.get().grant(options).then(() => setTimeout(f, 100), onFailure)
```
Is there a reason behind 100ms specifically?

breilly2 replied:
It's what I found to work experimentally. I tested with a 1ms delay and that wasn't long enough. I never got an error with 100ms, and that felt like an acceptable pause to introduce. Of course, I don't know for sure that 100ms will always be enough time. :(

breilly2 committed:
withBillingScope -> ensureBillingScope

panentheos commented on src/libs/ajax.js:
```diff
@@ -655,6 +658,34 @@ const Buckets = signal => ({
+const GoogleBilling = signal => ({
+  listAccounts: async () => {
+    if (!hasBillingScope()) {
+      reportError('Terra does not have access to query your Google Cloud Billing information')
```
We generally don't do error reporting at this level, but just let the error propagate out. That gives the consuming code more control over the user experience.

breilly2 replied:
Good call. I'll clean that up.

breilly2 replied:
@panentheos Do you think we should call `ensureBillingScope` right here?

panentheos replied:
Perhaps. But without a concrete use case it's hard to speculate. These are good building blocks, and as we go forward it will probably become clearer how best to assemble them.

panentheos commented on src/libs/auth.js:
```diff
@@ -21,6 +21,28 @@ export const reloadAuthToken = () => {
+  } else {
+    const options = new window.gapi.auth2.SigninOptionsBuilder({ 'scope': 'https://www.googleapis.com/auth/cloud-billing' })
+    // Wait 100ms before doing the thing to avoid errors due to delays in applying the new scope grant
+    return getAuthInstance().currentUser.get().grant(options).then(() => new Promise(resolve => setTimeout(resolve, 100)))
```
Since this function is `async`, we can leverage `await` instead of explicit promises. E.g.:

```
const ensureBillingScope = async () => {
  if (!hasBillingScope()) {
    await /* grant the scope */
    await Utils.delay(100)
  }
}
```

breilly2 replied:
TIL about `Utils.delay`.

breilly2 committed:
Let errors escape from front-line ajax calls.
Added billing.js with listProjectsWithAccounts function.
Simplified ensureBillingScope.

breilly2 commented:
I added a `listProjectsWithAccounts` function to the library as a strawman of how the pieces (which kept getting smaller and smaller) might be assembled. That function implementation is not set in stone and should be adapted to the use case needed for #1188.

katevoss commented on src/libs/auth.js:
```diff
@@ -21,6 +21,27 @@ export const reloadAuthToken = () => {
+    const options = new window.gapi.auth2.SigninOptionsBuilder({ 'scope': 'https://www.googleapis.com/auth/cloud-billing' })
+    await getAuthInstance().currentUser.get().grant(options)
+    // Wait 100ms before continuing to avoid errors due to delays in applying the new scope grant
+    await Utils.delay(100)
```
This should resolve #1331, nice!

breilly2 replied:
Unfortunately, I don't think so. This serves the same role as the 250ms delay from your earlier PR (before the original origin of this branch). If we're still seeing the error after waiting 250ms then we'll probably need something more sophisticated here, like polling the auth instance until it says the scope is granted instead of just waiting for a fixed time interval.

breilly2 committed:
Refactor granting of billing scope

breilly2 committed:
Fetch billing accounts from Rawls instead of directly from Google

katevoss commented on src/libs/auth.js:
```diff
@@ -21,6 +21,27 @@ export const reloadAuthToken = () => {
+  if (!hasBillingScope()) {
+    const options = new window.gapi.auth2.SigninOptionsBuilder({ 'scope': 'https://www.googleapis.com/auth/cloud-billing' })
+    await getAuthInstance().currentUser.get().grant(options)
+    // Wait 100ms before continuing to avoid errors due to delays in applying the new scope grant
```
Note that the number of ms in the comment is different than the actual delay in Line 41

breilly2 replied:
Fixed

katevoss reviewed (approved):
I tested this out by revoking then re-granting scopes, and it works as expected. Since it's mostly backend changes, this has little user-impact.

breilly2 committed:
Fix delay duration in comment
